controllers/topology/internal/extensions/patches/api/interface.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package api contains the API definition for the patch engine.
+// NOTE: We are introducing this API as a decoupling layer between the patch engine and the concrete components
+// responsible for generating patches, because we aim to provide support for external patches in a future iteration.
+// We also assume that this API and all the related types will be moved in a separate (versioned) package thus
+// providing a versioned contract between Cluster API and the components implementing external patch extensions.
+package api
+
+import (
+	"context"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+// Generator defines a component that can generate patches for ClusterClass templates.
+type Generator interface {
+	// Generate generates patches for templates.
+	// GenerateRequest contains templates and the corresponding variables.
+	// GenerateResponse contains the patches which should be applied to the templates of the GenerateRequest.
+	Generate(ctx context.Context, request *GenerateRequest) (*GenerateResponse, error)
+}
+
+// GenerateRequest defines the input for a Generate request.
+type GenerateRequest struct {
+	// Variables is a name/value map containing variables.
+	Variables map[string]apiextensionsv1.JSON
+
+	// Items contains the list of templates to generate patches for.
+	Items []*GenerateRequestTemplate
+}
```

Comment:
Should we add json tags to each one of these structs?

Reply:
I can do it. I'm not sure if we should already, given that we currently don't marshal/unmarshal those structs and we don't know if we need JSON tags, gRPC tags, both or maybe even define the structs in proto.

Reply:
That's fair, let's do it later